metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/ScalanetConnectionProvider.scala

```diff
@@ -0,0 +1,132 @@
+import java.security.SecureRandom
+
+object ScalanetConnectionProvider {
+  private class ScalanetEncryptedConnection[F[_]: TaskLift, K: Codec, M: Codec](
+      underlyingChannel: Channel[PeerInfo, M],
+      underlyingChannelRelease: F[Unit]
+  ) extends EncryptedConnection[F, K, M] {
+
+    override def close(): F[Unit] = underlyingChannelRelease
+
+    override lazy val remotePeerInfo: (K, InetSocketAddress) = (
+      Codec[K].decodeValue(underlyingChannel.to.id).require,
+      underlyingChannel.to.address.inetSocketAddress
+    )
+
+    override def sendMessage(m: M): F[Unit] =
+      TaskLift[F].apply(underlyingChannel.sendMessage(m))
+
+    override def incomingMessage: F[Option[Either[ConnectionError, M]]] = {
+      TaskLift[F].apply(underlyingChannel.nextChannelEvent.map {
+        case Some(event) =>
+          event match {
+            case Channel.MessageReceived(m) => Some(Right(m))
+            case Channel.UnexpectedError(e) => Some(Left(UnexpectedError(e)))
+            case Channel.DecodingError      => Some(Left(DecodingError))
+          }
+        case None => None
+      })
+    }
+  }
+  // Codec constraint for K is necessary as scalanet require peer key to be in BitVector format
+  def scalanetProvider[F[_]: Sync: TaskLift, K: Codec, M: Codec](
+      bindAddress: InetSocketAddress,
+      nodeKeyPair: AsymmetricCipherKeyPair,
+      secureRandom: SecureRandom,
+      useNativeTlsImplementation: Boolean,
+      framingConfig: FramingConfig,
+      maxIncomingQueueSizePerPeer: Int
+  )(implicit
+      sch: Scheduler
+  ): Resource[F, EncryptedConnectionProvider[F, K, M]] = {
+    for {
+      config <- Resource.liftF[F, Config](
+        Sync[F].fromTry(
+          DynamicTLSPeerGroup
+            .Config(
+              bindAddress,
+              Secp256k1,
+              nodeKeyPair,
+              secureRandom,
+              useNativeTlsImplementation,
+              framingConfig,
+              maxIncomingQueueSizePerPeer,
+              None
+            )
+        )
+      )
+      pg <- DynamicTLSPeerGroup[M](config).mapK(TaskLift.apply)
+    } yield new EncryptedConnectionProvider[F, K, M] {
+      override def localInfo: (K, InetSocketAddress) = (
+        Codec[K].decodeValue(pg.processAddress.id).require,
+        pg.processAddress.address.inetSocketAddress
+      )
+
+      override def connectTo(
+          k: K,
+          address: InetSocketAddress
+      ): F[EncryptedConnection[F, K, M]] = {
+        val encodedKey = Codec[K].encode(k).require
+        TaskLift[F].apply(
+          pg.client(PeerInfo(encodedKey, InetMultiAddress(address)))
+            .allocated
+            .attempt
+            .flatMap {
+              case Left(value) => Task.raiseError(value)
+              case Right((channel, release)) =>
+                Task.now(
+                  new ScalanetEncryptedConnection(
+                    channel,
+                    TaskLift[F].apply(release)
```

Comment:
actually, it won't be possible (or easy) at this level. As scalanet `Channel` interface does not expose channel local info, and to properly distinguish outgoing channels at this level we need to known their ephemeral ports, so either we should do this at `DynamicTLSPeerGroup` level (which probably makes sense) or expose info about ephemeral ports assigned to outgoing channels

Reply:
Hm, I just meant that if whatever resource handles the incoming/outgoing connections is released, it could make sure that all registered connections are also closed. Not sure why you'd need to know more about ports.

Reply:
But I see this is happening now in the tests :+1: